File: 2466-count-ways-to-build-good-strings/2466-count-ways-to-build-good-strings.js
/**
 * @param {number} low
 * @param {number} high
 * @param {number} zero
 * @param {number} one
 * @return {number}
 */

let lows, highs, zeros, ones, dp, mod = 1e9+7;

var solve = (len) => {
    if(len > highs) {
        return 0;
    }

    if(dp[len] != -1) {
        return dp[len];
    }

    let addOne = 0;
    if(len >= lows) {
        addOne = 1;
    }

    return dp[len] = (addOne + solve(len + zeros) % mod + solve(len + ones) % mod) % mod;
}

var countGoodStrings = function(low, high, zero, one) {
    lows = low, highs = high, zeros = zero, ones = one;
    dp = new Array(highs + 1).fill(-1);
    return solve(0)
};

File: 2466-count-ways-to-build-good-strings/2466-count-ways-to-build-good-strings.cpp
class Solution {
    int lows, highs, ones, zeros;
    const int mod = 1e9+7;
    vector<int> dp;
public:
    int solve(int length) {
        if(length > highs) return 0;
        if(dp[length] != -1) return dp[length];

        int addOne = 0;
        if(length >= lows && length <= highs) {
            addOne = 1;
        }

        return dp[length] = (addOne + solve(length + zeros) % mod + solve(length + ones) % mod) % mod;
    }
    int countGoodStrings(int low, int high, int zero, int one) {
        lows = low, highs = high, ones = one, zeros = zero;

        dp.assign(highs + 1, -1);
        
        return solve(0);
    }
};